### PR TITLE
feat(tweet): SCRUM-57 Add Tweet Updating API

### PR DESCRIPTION
**Pull request by adnlv**
**Jira**: [SCRUM-57](https://epam-team-php.atlassian.net/browse/SCRUM-57)

## Description

Provide a protected tweet updating API.

## How to roll back?

Revert this pull request.

## How has this been tested?

Unit tests cover both the domain and application layers.

To reproduce, run:

```shell
docker compose exec php ./vendor/bin/phpunit tests/Tweet --display-all-issues --testdox --group unit --columns=40
```

You should see the following output with all tests passed:

```
PHPUnit 12.5.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.4.16
Configuration: /app/phpunit.dist.xml

................          16 / 16 (100%)

Time: 00:00.013, Memory: 16.00 MB

Create Tweet Command Handler (Twitter\Tests\Tweet\Application\UseCase\CreateTweet\CreateTweetCommandHandler)
 ✔ Handle creates tweet with provided data
 ✔ Handle propagates user not found exception

Tweet (Twitter\Tests\Tweet\Domain\Tweet\Model\Tweet)
 ✔ Create succeeds with valid data
 ✔ Create fails with invalid user id with data set "empty string"
 ✔ Create fails with invalid user id with data set "whitespace only"
 ✔ Create fails with invalid user id with data set "random string"
 ✔ Create fails with invalid content with data set "empty string"
 ✔ Create fails with invalid content with data set "whitespace only"
 ✔ Create fails with invalid content with data set "too long content"

Tweet Update Content (Twitter\Tests\Tweet\Domain\Tweet\Model\TweetUpdateContent)
 ✔ Updates content and refreshes updated at
 ✔ Throws exception when content is not valid with data set "empty content"
 ✔ Throws exception when content is not valid with data set "whitespace only"
 ✔ Throws exception when content is not valid with data set "too long content"

Update Tweet Command Handler (Twitter\Tests\Tweet\Application\UseCase\UpdateTweet\UpdateTweetCommandHandler)
 ✔ Updates tweet for owner
 ✔ Throws exception when tweet is updated not by the owner
 ✔ Throws exception when tweet does not exist

OK (16 tests, 30 assertions)
```

### Update an existing tweet

image

image

### Access token has expired

Screenshot 2026-01-06 093326

### Access denied

image

### CODE REVIEW

**Review by CPAprince — APPROVED** [review]
Good

File: composer.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
     },
     "autoload": {
         "psr-4": {
-            "Twitter\\": "src/"
+            "Twitter\\": "src/",
+            "App\\": "src/"
```
Comment (ihpr) [inline composer.json]: @adnlv Why are we keeping both?

File: src/Shared/Infrastructure/EventSubscriber/ExceptionSubscriber.php [hunk 2]
```diff
@@ -70,6 +71,11 @@
             'message' => 'The user with this ID was not found',
             'status' => Response::HTTP_NOT_FOUND,
         ],
+        TweetAccessDeniedException::class => [
+            'code' => 'ACCESS_DENIED',
+            'message' => 'You are not allowed to access this tweet',
+            'status' => Response::HTTP_FORBIDDEN,
```
Comment (ihpr) [inline src/Shared/Infrastructure/EventSubscriber/ExceptionSubscriber.php:77]: @adnlv (low) No need to use both classes `JsonResponse` and `Response` to get access to status codes. Leave only one.

File: tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * @throws TweetAccessDeniedException
+     */
+    #[Test]
+    public function propagatesTweetNotFoundException(): void
+    {
+        $this->tweetRepository
+            ->expects(self::once())
+            ->method('getById')
+            ->willThrowException(new TweetNotFoundException('non-existent-id'));
```
Comment (ihpr) [inline tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php:110]: @adnlv (high) Does it make sense to test exception propagation here? What if we test `tweetNotFound` behaviour instead?

File: tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        self::assertSame('New content', $result->content);
+        self::assertSame('New content', $tweet->content());
+        self::assertInstanceOf(DateTimeImmutable::class, $result->updatedAt);
+    }
+
+    /**
+     * @throws TweetNotFoundException
+     */
+    #[Test]
+    public function throwsAccessDeniedWhenUserIsNotOwner(): void
```
Comment (ihpr) [inline tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php]: @adnlv (low) I think the test doens't express the behaviour well enough. How about this name: `allowTweetUpdateOnlyForTheOwner` or `throwsExceptionWhenTweetIsUpdatedNotByTheOwner`?

File: tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        $this->tweetRepository = $this->createMock(TweetRepository::class);
+        $this->handler = new UpdateTweetCommandHandler($this->tweetRepository);
+    }
+
+    /**
+     * @throws TweetAccessDeniedException
+     * @throws TweetNotFoundException
+     */
+    #[Test]
+    public function updatesTweetWhenUserIsOwner(): void
```
Comment (ihpr) [inline tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php]: @adnlv (low) I think the name doesn't express the tested behaviour well enough. I see no test for user specific role. How about: `shouldUpdateTweet`?

File: tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        $tweet = Tweet::create(
+            '123e4567-e89b-12d3-a456-426614174000',
```
Comment (ihpr) [inline tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php:41]: @adnlv (medium) Is that the only correct data set?

File: tests/Tweet/Domain/Tweet/Model/TweetUpdateContentTest.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            'Initial valid content',
+        );
+
+        $this->expectException(LazyAssertionException::class);
+
+        $tweet->updateContent('');
+    }
+
+    #[Test]
+    public function throwsExceptionWhenContentExceedsLengthLimit(): void
```
Comment (ihpr) [inline tests/Tweet/Domain/Tweet/Model/TweetUpdateContentTest.php]: @adnlv (medium) I see that this test is a complete copy of the test above. I mean, the test behaviour we are testing here is the same. You're trying to pass invalid content.

How do you think it would be better to keep one test for a single behaviour, but create different datasets?

File: tests/Tweet/Domain/Tweet/Model/TweetUpdateContentTest.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+use PHPUnit\Framework\Attributes\Test;
+use PHPUnit\Framework\TestCase;
+use Twitter\Tweet\Domain\Tweet\Model\Tweet;
+
+#[Group('unit')]
+#[CoversMethod(Tweet::class, 'updateContent')]
+final class TweetUpdateContentTest extends TestCase
+{
+    #[Test]
+    public function updatesContentAndUpdatedAt(): void
```
Comment (ihpr) [inline tests/Tweet/Domain/Tweet/Model/TweetUpdateContentTest.php]: @adnlv (medium) based on the name, I see an attempt to test 2 diferent behaviour here:
1. `updatesContent`
2. `keepUpdatedAtRefreshed`

Would it be better to split those?

**Review by ihpr — DISMISSED** [review]

File: src/Shared/Infrastructure/EventSubscriber/ExceptionSubscriber.php (hunk 2 above)
Comment (adnlv) [reply src/Shared/Infrastructure/EventSubscriber/ExceptionSubscriber.php:77]: @ihpr, please read my [previous response](https://github.com/CPAprince/Twitter/pull/29#discussion_r2662333238).

File: tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php (hunk 6 above)
Comment (adnlv) [reply tests/Tweet/Application/UseCase/UpdateTweet/UpdateTweetCommandHandlerTest.php:41]: @ihpr, the specific values here are not significant for the behavior under test. The goal is to set up the minimal valid state required to exercise the happy path of the handler. Since the logic does not branch on the UUID or content value beyond validity and equality, varying the dataset would not provide additional coverage and would only add noise. Edge cases and validation are covered by separate tests at the appropriate level.

File: tests/Tweet/Domain/Tweet/Model/TweetUpdateContentTest.php (hunk 8 above)
Comment (adnlv) [reply tests/Tweet/Domain/Tweet/Model/TweetUpdateContentTest.php]: @ihpr, I’m keeping the assertions together because they validate a single domain operation (`updateContent`) and its contractual side effects. Updating the content without refreshing `updatedAt` would be an invalid state, so separating them would weaken the invariant rather than clarify it.
